mock/py/mockbuild/plugins/overlayfs.py

```diff
@@ -0,0 +1,815 @@
+        dataBaseDir = self.getPluginInstanceDir()
+        if not os.path.exists(dataBaseDir):
+            os.mkdir(dataBaseDir)
+        layersDir = self.getLayersDir()
+        if not os.path.exists(layersDir):
+            os.mkdir(layersDir)
+        refsDir = self.getRefsDir()
+        if not os.path.exists(refsDir):
+            os.mkdir(refsDir)
+        locksDir = self.getLocksDir()
+        if not os.path.exists(locksDir):
+            os.mkdir(locksDir)
+
+    # init basic refs/layers setup (create special layers/refs)
+    def initLayers(self):
+        baseLayerRef = self.getBaseLayerRef()
+        if not self.refExists(baseLayerRef):
+            self.createLayerAndRef(baseLayerRef, None)
+            self.setLayerImmutable(self.getLayerFromRef(baseLayerRef))
+        upperLayerRef = self.getUpperLayerRef()
+        if not self.refExists(upperLayerRef):
+            self.createRef(upperLayerRef, self.getLayerFromRef(baseLayerRef))
+        currentLayerRef = self.getCurrentLayerRef()
+        if not self.refExists(currentLayerRef):
+            self.createRef(currentLayerRef, self.getLayerFromRef(baseLayerRef))
+
+    # this makes sure nothing is written to snapshot layers
+    # ( once snapshot is done it's layer becomes read only )
+    def prepareLayersForMount(self):
+        upperLayerRef = self.getUpperLayerRef()
+        upperLayer = self.getLayerFromRef(upperLayerRef)
+        # if upperLayerRef points to layer, which is marked immutable
+        # we cannot use that layer as upper layer, we need to create new one
+        # which has current upperLayer as parent and set it as upperLayer
+        if self.isLayerImmutable(upperLayer):
+            newLayerId = self.createLayer(upperLayer)
+            self.deleteRef(upperLayerRef)
+            self.createRef(upperLayerRef, newLayerId)
+
+    # create list of layer and all its parents
+    # (used when mounting it as overlayfs)
+    def createLayerList(self, layerId):
+        layerList = []
+        self.createLayerList2(layerList,layerId)
+        return layerList
+
+    def createLayerList2(self, layerList,layerId):
+        parentLayerId = self.getParentLayer(layerId)
+        layerList.append(layerId)
+        if parentLayerId != None:
+            self.createLayerList2(layerList,parentLayerId)
+
+    # mount root: upperLayer (+ its parents) using overlayfs
+    def mountRoot(self):
+        self.prepareLayersForMount()
+
+        upperLayerRef = self.getUpperLayerRef()
+        upperLayerId = self.getLayerFromRef(upperLayerRef)
+
+        lowerTopLayerId = self.getParentLayer(upperLayerId)
+        lowerList = self.createLayerList(lowerTopLayerId)
+
+        workDir = self.getWorkDir()
+        if os.path.exists(workDir):
+            shutil.rmtree(workDir)
+        os.mkdir(workDir)
+
+        # make sure kernel has required module loaded
+        modprobeCmds = []
+        modprobeCmds.append("modprobe")
+        modprobeCmds.append("overlay")
```

Comment:
Can be done in one step, which is more effecient `modprobeCmds = ["modprobe", "overlay"]`